livekit/src/room/participant/local_participant.rs

```diff
@@ -544,4 +657,194 @@ impl LocalParticipant {
+            Err(_) => {
+                self.local.pending_responses.lock().remove(&id);
+                return Err(RpcError::built_in(RpcErrorCode::ResponseTimeout, None));
+            }
+            Ok(result) => result,
+        };
+
+        match response {
+            Err(_) => {
+                // Something went wrong locally
+                Err(RpcError::built_in(RpcErrorCode::RecipientDisconnected, None))
+            }
+            Ok(Err(e)) => {
+                // RPC error from remote, forward it
+                Err(e)
+            }
+            Ok(Ok(payload)) => {
+                // Successful response
+                Ok(payload)
+            }
+        }
+    }
+
+    pub fn register_rpc_method(
+        &self,
+        method: String,
+        handler: impl Fn(
+                String,
+                ParticipantIdentity,
+                String,
+                Duration,
+            ) -> Pin<Box<dyn Future<Output = Result<String, RpcError>> + Send>>
+            + Send
+            + Sync
+            + 'static,
+    ) {
+        self.local.rpc_handlers.lock().insert(method, Arc::new(handler));
+    }
+
+    pub fn unregister_rpc_method(&self, method: String) {
+        self.local.rpc_handlers.lock().remove(&method);
+    }
+
+    pub(crate) fn handle_incoming_rpc_ack(&self, request_id: String) {
+        if let Some(tx) = self.local.pending_acks.lock().remove(&request_id) {
+            let _ = tx.send(());
+        } else {
+            log::error!("Ack received for unexpected RPC request: {}", request_id);
+        }
+    }
+
+    pub(crate) fn handle_incoming_rpc_response(
+        &self,
+        request_id: String,
+        payload: Option<String>,
+        error: Option<RpcError_Proto>,
+    ) {
+        if let Some(tx) = self.local.pending_responses.lock().remove(&request_id) {
+            let _ = tx.send(match error {
+                Some(e) => Err(RpcError::from_proto(e)),
+                None => Ok(payload.unwrap_or_default()),
+            });
+        } else {
+            log::error!("Response received for unexpected RPC request: {}", request_id);
+        }
+    }
+
+    pub(crate) async fn handle_incoming_rpc_request(
+        &self,
+        caller_identity: ParticipantIdentity,
+        request_id: String,
+        method: String,
+        payload: String,
+        response_timeout_ms: u32,
+    ) {
+        if let Err(e) = self.publish_rpc_ack(caller_identity.to_string(), request_id.clone()).await
+        {
+            log::error!("Failed to publish RPC ACK: {:?}", e);
+        }
+
+        let handler = self.local.rpc_handlers.lock().get(&method).cloned();
+
+        let caller_identity_2 = caller_identity.clone();
+        let request_id_2 = request_id.clone();
+
+        let response = match handler {
+            Some(handler) => {
+                match tokio::task::spawn(async move {
+                    match panic::catch_unwind(AssertUnwindSafe(|| {
```

Comment:
mmh we shouldn't use a panic handler here, it is too opiniated

Reply:
Rust doesn't raise exceptions, but rather we return Results

Reply:
I think it's important to ensure that the caller on the other side of the RPC gets notified when an error occurs at the destination that prevents a normal response from being returned (similar to an HTTP 500 error). To achieve that, the framework must catch any kind of unhandled error in the response handler so that it can generate the error response (called `1500 ApplicationError` in the RPC framework). Maybe we should re-throw the exception afterwards so the developer can still collect them? Of course if they have their own error handling try/catch construct within their handler, then they can catch and return any error response they'd like. This all applies only to uncaught errors coming from their handler.

That's the language agnostic above - it will differ a bit per platform - but in Rust I think this means we should be catching panics. If we wanted to implement "rethrow" then my understanding is we'd use `resume_unwind` in Rust but I don't have experience with it.

What do you think?

Reply:
yeah the handler already is `Result<String, RpcError>`. That lets the user send custom errors through (you want them to be able to send specific errors but also you want to turn errors they didn't intend into generic errors so you don't leak private info across the channel).

take a look at the example, which [has a test case for this](https://github.com/livekit/rust-sdks/pull/461/files#diff-9584e9c9d2b409f287a1e1703a3760bb2a7b557119a86a4571c5dca6aef5f9f3R133). in this example, there is a division function exposed to RPC. but the caller sends 0 in the divisor. because the division function is poorly implemented and doesn't check for that, it attempts the divide by zero and causes a panic. this seems like a pretty normal thing to mess up - i'm fine with letting the local program crash and the dev clean it up, but I don't like the idea of the remote caller being stuck waiting for a response that will never come.

Reply:
ok so i just went back over this and found that catch_unwind was unnecessary to satisfy the example. the existing tokio::task::spawn behavior was already doing it and returning an Err.  simplified